bootloader/src/pyi_utils.c

```diff
@@ -1076,185 +1076,243 @@ static const char *CC2Str(FourCharCode code) {
+               (UInt32)actualType, CC2Str((FourCharCode)actualType), (long)actualSize);
+            VS("LOADER [AppleEvent EVT_FWD]: Putting param.\n");
+            err = AEPutParamPtr(&evtCopy, keyDirectObject, actualType, buf, actualSize);
+            if (err != noErr) goto cleanup1;
         }
-    } else {
-        /* Not GetURL or OpenDoc, so just ignore */
-        VS("LOADER [AppleEvent]: Handler ignoring event '%s'\n", CC2Str(evtCode));
+        VS("LOADER [AppleEvent EVT_FWD]: Sending message...\n");
+        err = AESendMessage(&evtCopy, NULL, kAENoReply, 60 /* 60 = about 1.0 seconds timeout */);
+        VS("LOADER [AppleEvent EVT_FWD]: Handler forwarded %s message to child pid %ld.\n",
+           descStr, (long)child_pid);
+
+    cleanup1:
+        free(buf); /* free of possible NULL ptr ok */
+        AEDisposeDesc(&evtCopy); /* dispose apple event copy */
+    cleanup2:
+        AEDisposeDesc(&target);
+        if (err)
+            OTHERERROR("LOADER [AppleEvent EVT_FWD]: %s handler got error %d\n", descStr, (int)err);
+        return (OSErr)err;
+    }
+    return noErr; /* not reached */
+}
+
+/* Handler for 'rapp' (reopen app) events. This brings the child_pid's windows
+ * to the foreground when the user double-clicks the app's icon again in the macOS UI.
+ * 'rapp' is sent to us only when the app is already running. */
+static OSErr handle_rapp_events(void)
+{
+    OSErr err;
+    ProcessSerialNumber psn;
+
+    VS("LOADER [AppleEvent]: ReopenApp handler called.\n");
+    if (!child_pid) {
+        /* Child not up yet -- there is no way to "forward" this before child started!
+         * It's safe to ignore this event now since the subordinate application will
+         * activate anyway after it started. */
+         VS("LOADER [AppleEvent]: Ignoring spurious ReopenApp event (child_pid is 0)\n");
+         return errAEEventNotHandled;
+    }
+    err = GetProcessForPID(child_pid, &psn);
+    if (err != noErr) {
+         OTHERERROR("LOADER [AppleEvent]: Failed to get psn for pid %d, maybe the child process just exited?\n",
+                    (int)child_pid);
+         return err;
+    }
+    VS("LOADER [AppleEvent]: Raising process and returning.\n");
+    /* SetFrontProcess() below is deprecated -- but it is the only way we have to do this.
+     * The below noisy #ifdef block basically disables the warning on clang. */
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+    /* We cannot actually forward the 'rapp' message to a process by ProcessSerialNumber.
+     * We can only do it to a bundle-id.  Since the subordinate app has no real bundle-id,
+     * we must instead "simulate" the "rapp" message by doing this macOS call which sends
+     * the right combination of events to the subordinate app to get it to active and raise
+     * itself to the foreground. */
+    err = SetFrontProcess(&psn);
```

Comment:
Hmm it is quite an old deprecated function... I can help to find an alternative or are you 100% sure there isn't?

Reply:
There is if you use Objective-C -- in straight C.. no.  They deprecated these ancient C APIs in favor of their ObjC stuff (and also Swift which is the same API really).  I don't believe they will ever remove this API call since it would break old 
apps to do so.  There are even more ancient calls they have never removed...

I would like very much to remove this call but I fear without it users may not get the behavior they expect.

The thing is without this call the subordinate app doesn't behave as we would expect.  I suspect there are more events that we need to intercept to get the full behavior -- and I just don't know what they are.

Using this call basically makes the observed behavior work as if pyinstaller were not there.  Without this -- it doesn't do what we want.

I can also figure out if there is another way .. it may involve intercepting more event types.. and the documentation on this is sparse...

Reply:
If anyone reading this knows what other event types I would need to intercept to get the expected behavior of:

1. reopen app when already running (as with `open App.app`) 
2. all app windows pop to front

It appears simply forwarding `rapp` is not enough...

Reply:
Hmm let me try a few more things. I'll get back to you.

Reply:
Update: Ok.  I got rid of the deprecated call and hack.  I got to the bottom of it. See description.

Reply:
You did it 💪 !

Reply:
Yeah -- feels good to do it right and not in a cheap way. :)  Also I believe this is now the correct order of events to the child app which hopefully means fewer support tickets later... since I think now we are respecting whatever contract the child GUI framework might expect.